### PR TITLE
Features/different network

### PR DESCRIPTION
**Pull request by karthikeyan-cere**
Switch network for friendly bot.

### CODE REVIEW

File: src/modules/friendly-bot/entities/payout.entity.ts [hunk 1]
```diff
@@ -1,5 +1,10 @@
 import {Column, CreateDateColumn, Entity, PrimaryGeneratedColumn} from 'typeorm';
 
+export enum NetworkEnum {
+  TESTNET = '0',
```
Comment (evgeny-s) [inline src/modules/friendly-bot/entities/payout.entity.ts:4]: Can we use the same name like: TESTNET='TESTNET'

File: src/modules/friendly-bot/friendly-bot.interface.ts [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import {AssetDto} from './dto/assets.dto';
 
 export interface FriendlyBotServiceInterface {
-  issueToken(destination: string): Promise<AssetDto>;
+  issueToken(destination: string, network:string): Promise<AssetDto>;
```
Comment (evgeny-s) [inline src/modules/friendly-bot/friendly-bot.interface.ts:4]: Let's use enum for the type here.

File: src/modules/friendly-bot/friendly-bot.service.ts [hunk 3]
```diff
@@ -7,50 +7,105 @@ import {ApiPromise, Keyring, WsProvider} from '@polkadot/api';
 import {KeypairType} from '@polkadot/util-crypto/types';
 import {KeyringPair} from '@polkadot/keyring/types';
 import moment from 'moment';
-import {PayoutEntity} from './entities/payout.entity';
+import {PayoutEntity, NetworkEnum} from './entities/payout.entity';
 import {AssetDto} from './dto/assets.dto';
 import {BalanceDto} from './dto/balance.dto';
 import {Hash} from '@polkadot/types/interfaces';
 
+enum NETWORKS {
```
Comment (evgeny-s) [inline src/modules/friendly-bot/friendly-bot.service.ts:15]: Why do we have duplication for Networks?
Remove from here, move the separate file

File: src/modules/friendly-bot/friendly-bot.service.ts [hunk 4]
```diff
@@ -7,50 +7,105 @@ import {ApiPromise, Keyring, WsProvider} from '@polkadot/api';
...
+  testNet = 'TESTNET',
+  dev = 'TESTNET_DEV',
+  dev1 = 'TESTNET_DEV1',
+}
 @Injectable()
 export class FriendlyBotService implements FriendlyBotServiceInterface {
   public logger = new Logger(FriendlyBotService.name);
 
-  private api: ApiPromise;
+  private testNetApi: ApiPromise;
```
Comment (evgeny-s) [inline src/modules/friendly-bot/friendly-bot.service.ts:24]: Let's do it in more flexible way:
```
private networks: { api: ApiPromise, keyRing: KeyringPair }[] = []; // custom object can be also moved to DTO
```

File: src/modules/friendly-bot/friendly-bot.service.ts [hunk 5]
```diff
@@ -7,50 +7,105 @@ import {ApiPromise, Keyring, WsProvider} from '@polkadot/api';
...
     @InjectRepository(PayoutEntity)
     private readonly botEntityRepository: Repository<PayoutEntity>,
   ) {
-    this.init();
+    this.testNetInit();
+    this.devInit();
+    this.dev1Init();
+  }
+
+  public async testNetInit(): Promise<ApiPromise> {
```
Comment (evgeny-s) [inline src/modules/friendly-bot/friendly-bot.service.ts:48]: Combine everything in 1 method.
The idea - we should be able to add new network, just by adding parameters to the ENV file

File: src/modules/friendly-bot/friendly-bot.service.ts [hunk 6]
```diff
@@ -64,43 +119,97 @@ export class FriendlyBotService implements FriendlyBotServiceInterface {
...
+        botEntity.network = NetworkEnum.DEV;
+        break;
+      }
+      case NETWORKS.dev1: {
+        botEntity.sender = this.dev1appKeyring.address;
+        botEntity.network = NetworkEnum.DEV1;
+        break;
+      }
+      default:
+        break;
```
Comment (evgeny-s) [inline src/modules/friendly-bot/friendly-bot.service.ts:153]: Throw error here about incorrect network parameter

File: src/modules/friendly-bot/friendly-bot.service.ts [hunk 7]
```diff
@@ -64,43 +119,97 @@ export class FriendlyBotService implements FriendlyBotServiceInterface {
...
+      }
+      case NETWORKS.dev1: {
+        const account = await this.testNetApi.query.system.account(address);
+        balance = account.data.free.toString();
+        decimal = await this.dev1Api.registry.chainDecimals;
+        break;
+      }
+      default:
+        break;
```
Comment (evgeny-s) [inline src/modules/friendly-bot/friendly-bot.service.ts:184]: Throw error here about incorrect network parameter

File: .env.local [hunk 8]
```diff
@@ -4,17 +4,28 @@ DB_PASSWORD=block-scanner-service
...
 PORT=2015
 
 LOG_LEVEL=debug
 
-NETWORK_WS_URL=wss://testnet-node-1.cere.network:9945
-SUBSTRATE_APP_WALLET_JSON=SUBSTRATE_APP_WALLET_JSON
-SUBSTRATE_APP_WALLET_PASSPHRASE=SUBSTRATE_APP_WALLET_PASSPHRASE
+NETWORK_WS_URL=NETWORKNET_WS_URL
+
+TESTNET_WS_URL=TESTNET_WS_URL
```
Comment (evgeny-s) [inline .env.local:17]: Can we convert it to array of objects?
```
NETWORK_PARAMS=[{url: '', wallet: '', passphrase: ''}]
```

**Review by evgeny-s — CHANGES_REQUESTED** [review]
Please check comments and don't forget to fix merge conflicts before submitting for review.

File: src/modules/friendly-bot/friendly-bot.service.ts [hunk 9]
```diff
@@ -64,43 +78,56 @@ export class FriendlyBotService implements FriendlyBotServiceInterface {
...
-    const hash = (await this.transfer(destination, value)).toString();
+    const hash = (await this.transfer(destination, value, network)).toString();
     const botEntity = new PayoutEntity();
     botEntity.destination = destination;
-    botEntity.sender = this.appKeyring.address;
     botEntity.txnHash = hash;
     botEntity.value = value;
+    botEntity.sender = this.appKeyring.address;
+    switch (network) {
+      case NetworkEnum.TESTNET: {
```
Comment (evgeny-s) [inline src/modules/friendly-bot/friendly-bot.service.ts:96]: Can we remove this switch/case and get parameter form function argument?

File: src/modules/friendly-bot/entities/payout.entity.ts (hunk 1 above)
Comment (karthikeyan-cere) [reply src/modules/friendly-bot/entities/payout.entity.ts]: Done.

File: src/modules/friendly-bot/friendly-bot.interface.ts (hunk 2 above)
Comment (karthikeyan-cere) [reply src/modules/friendly-bot/friendly-bot.interface.ts:4]: Done.

File: src/modules/friendly-bot/friendly-bot.service.ts (hunk 3 above)
Comment (karthikeyan-cere) [reply src/modules/friendly-bot/friendly-bot.service.ts]: Done.

File: src/modules/friendly-bot/friendly-bot.service.ts (hunk 4 above)
Comment (karthikeyan-cere) [reply src/modules/friendly-bot/friendly-bot.service.ts]: Done.

File: src/modules/friendly-bot/friendly-bot.service.ts (hunk 5 above)
Comment (karthikeyan-cere) [reply src/modules/friendly-bot/friendly-bot.service.ts]: Done.

File: src/modules/friendly-bot/friendly-bot.service.ts (hunk 6 above)
Comment (karthikeyan-cere) [reply src/modules/friendly-bot/friendly-bot.service.ts]: Done.

File: .env.local (hunk 8 above)
Comment (karthikeyan-cere) [reply .env.local]: Done

File: .env.local [hunk 10]
```diff
@@ -4,17 +4,22 @@ DB_PASSWORD=block-scanner-service
...
 LOG_LEVEL=debug
 
-NETWORK_WS_URL=wss://testnet-node-1.cere.network:9945
-SUBSTRATE_APP_WALLET_JSON=SUBSTRATE_APP_WALLET_JSON
-SUBSTRATE_APP_WALLET_PASSPHRASE=SUBSTRATE_APP_WALLET_PASSPHRASE
+NETWORK_WS_URL=NETWORK_WS_URL
+
+NETWORKS = [{"type": "TESTNET", "URL": "wss://testnet-node-1.cere.network:9945", "NETWORK": "TESTNET"},{"type": "TESTNET_DEV", "URL": "wss://testnet-node-1.cere.network:9945",  "NETWORK": "TESTNET_DEV"}, {"type": "TESTNET_DEV1", "URL": "wss://testnet-node-1.cere.network:9945", "NETWORK": "TESTNET_DEV1"} ]
+
+APP_WALLET_JSON=APP_WALLET_JSON
```
Comment (evgeny-s) [inline .env.local:19]: Let's put these parameters to the NETWORK, in case we want to change only for mainnet later. So we can configure them individually.

Comment (evgeny-s) [reply .env.local:19]: And of course - remove these parameteres from here. So we have only NETWORKS with such kind of configuration

File: src/modules/friendly-bot/entities/payout.entity.ts [hunk 11]
```diff
@@ -1,5 +1,10 @@
 import {Column, CreateDateColumn, Entity, PrimaryGeneratedColumn} from 'typeorm';
 
+export enum NetworkEnum {
```
Comment (evgeny-s) [inline src/modules/friendly-bot/entities/payout.entity.ts:3]: Let's remove this enum and use `string` type. In this case we can configure new networks only via ENV.

File: src/modules/friendly-bot/friendly-bot.interface.ts (hunk 2 above)
Comment (evgeny-s) [reply src/modules/friendly-bot/friendly-bot.interface.ts:4]: ignore this

File: src/modules/friendly-bot/friendly-bot.service.ts [hunk 12]
```diff
@@ -30,27 +30,41 @@ export class FriendlyBotService implements FriendlyBotServiceInterface {
...
+    await api.isReady;
+    const chain = await api.rpc.system.chain();
     this.logger.log(`Connected to ${chain}`);
-    const appWalletJson = this.configService.get('SUBSTRATE_APP_WALLET_JSON');
-    const appPassphrase = this.configService.get('SUBSTRATE_APP_WALLET_PASSPHRASE');
+    return api;
+  }
+
+  public initFaucet() {
+    const appWalletJson = this.configService.get('APP_WALLET_JSON');
```
Comment (evgeny-s) [inline src/modules/friendly-bot/friendly-bot.service.ts:52]: Remove this from here and initialize N different assets, based on network config.

File: src/modules/friendly-bot/friendly-bot.service.ts [hunk 13]
```diff
@@ -16,7 +16,7 @@ import {Hash} from '@polkadot/types/interfaces';
 export class FriendlyBotService implements FriendlyBotServiceInterface {
   public logger = new Logger(FriendlyBotService.name);
 
-  private api: ApiPromise;
+  private networkParams: {api: ApiPromise; type: string}[] = [];
 
   private keyRingType: KeypairType;
```
Comment (evgeny-s) [inline src/modules/friendly-bot/friendly-bot.service.ts:22]: In this case we don't need this property here.

**Review by evgeny-s — CHANGES_REQUESTED** [review]

File: src/modules/friendly-bot/friendly-bot.service.ts (hunk 9 above)
Comment (karthikeyan-cere) [reply src/modules/friendly-bot/friendly-bot.service.ts:96]: done.

Comment (karthikeyan-cere) [reply src/modules/friendly-bot/friendly-bot.service.ts:96]: done

File: .env.local (hunk 10 above)
Comment (karthikeyan-cere) [reply .env.local:19]: done.

File: src/modules/friendly-bot/entities/payout.entity.ts (hunk 11 above)
Comment (karthikeyan-cere) [reply src/modules/friendly-bot/entities/payout.entity.ts]: done

File: src/modules/friendly-bot/friendly-bot.service.ts (hunk 12 above)
Comment (karthikeyan-cere) [reply src/modules/friendly-bot/friendly-bot.service.ts]: done

File: src/modules/friendly-bot/friendly-bot.service.ts (hunk 13 above)
Comment (karthikeyan-cere) [reply src/modules/friendly-bot/friendly-bot.service.ts:22]: done

**Review by evgeny-s — APPROVED** [review]

**Pull request merged**